### ISSUE TITLE
Types of `gate` parameters, supported operations

### ISSUE THREAD

**Issue by levbishop** [openqasm/openqasm#375]
As a result of the work #346 we surfaced the pre-existing ambiguity about the type for the parameters to `gate` declarations. The  purpose of `gate` is to define mathematical equivalences between gates. Having imprecise types (floating/fixed point) interferes with this. If the parameters have some new "symbolic" type there's a question of what set of operations we should support on symbolic objects, and some thought that maybe linear combinations with rational coefficients may be sufficient to capture the relevant use cases (and allow us to stop short of requiring a full Computer Algebra system within openqasm).

The purpose of this issue is to pull out this parameter discussion from #346 and allow #346 to focus on the `gate` extension question.

**Comment by levbishop** [comment]
After much discussion in the T&C WG I'd like to put forward the following proposal. It ended up as a long essay so here's the outline.
1. `gate` just associates every call with an explicit unitary and gives a forward definition in terms of more fundamental operations. This supports both simple execution engines and also more sophisticated circuit rewriting, without putting any requirement on what kind of rewriting must be available.
2. The parameters of gates overflow/wrap exactly the same way that regular openqasm `angle`s do, this requires care but no more so than anywhere else `angle` types are used in openqasm.
3. There is an unfortunate inconsistency in the spec around the theta parameter of the U gate, and we have to pick one of a few possible ways to resolve the inconsistency
4. We need to choose a number of bits of precision for any manipulations of gate parameters.

### Part 1: Purpose of the `gate` statement, expectations on compilers
`gate` does 2 things: 1) associate a unique unitary with every call to the defined gate; 2) describe one concrete implementation of that gate in terms of other gates (after repeated unwrapping eventually in terms of fundamental gates). 

Compilers at minimum MUST be able to unroll recursively until reaching a set of gates for which `defcal`s are available, and MAY throw an error if any gates unroll all the way to the built-in `U` basis without hitting available defcals. (We expect that the profiles WG will define an "executable QASM" profile where erroring is appropriate, and 

If a vendor provides a standard gate library include file (eg stgates.inc) for a given execution environment, and user code builds only on gates from that library, then the user code should be able to execute (modulo other vendor constraints such as a maximum circuit depth limit etc). A simple minimal implementation (may not be very efficient) would therefore be to simply provide `defcal` for `U` and `CX` and standard libraries that build only on those (ie without further use of `ctrl` modifiers and such). (We should consider distributing such a `stdgates.inc` implementation along with the reference implemention - the example on github is not of this form). More efficient implementations of the standard libraries might unroll directly to other native gates.

If a compiler supports basis-changing the compiler MAY recognize some subset of `gate` definitions and have functionality to eg rewrite from an `crz(x)` basis to an `partialswap(x)` basis. This might require the `gate partialswap(x)` definition to be provided in a particular form because although the mapping is unique in the sense `partialswap(x)` describes a unique unitary, there may be multiple implementations of the body of `partialswap` and the compiler might only recognize some of them. When searching for known basis gates, the compiler must ignore the name of the gate and parameters (ie, `gate partialswap(x)` and `gate fractional_swap(theta)` should both be recognized if the body of the gate is identical after parameter renaming).

In this case that a compiler supports basis-changing, it MAY recognize as equivalent gates with syntactically-different but mathematically-equivalent definitions. For non-parameterized gates its probably practical to recognize all definitions of the same unitary (and there are examples of this in eg qiskit-terra), for parameterized gates it likely isn't practical to recognize all formulations but maybe simple transformations can be spotted.

I think of this model by analogy to eg vector acceleration instructions for classical CPUs. By providing a `defcal` for an instruction, it becomes available for direct use by the programmer, analogous to inline assembly to use a vector instruction. Additionally, some compilers may optionally, provide auto-vectorization functionality even when they do its expected that a compiler will require specific support for each such instruction set (eg just because SSE is supported doesnt mean SSE2 will be without additional compiler development) and its not expected that a classical compiler recognizes all vectorizable code, but rather the input should stick to certain idioms (eg side-effect-free loops over memory-aligned arrays).

### part 2: the wrapping behaviour of the `gate` parameters
There was a lot of discussion in the WG around how exactly the wrapping behavior of an `angle` type in a gate parameter should be implemented. This comes from the fact that eg `3x/2` when x is an `angle` can have very different result if its interpreted as `(3*x)/2` or `3*(x/2)`. Even though the spec takes pains to carefully describe the behaviour of such explicit expressions, the concern was that hiding some of this in `gate` definitions might make it implicit and mysterious.  I think with the above expectations on `gate` definitions this is no longer a problem and we can make a simple rule:
The overflow behaviour of recursively unrolling `gates` should be as-if the expression were evaluated by assigning to an angle type at each call.

Ie, with the definitions
```
gate g1(x) q {
    g2(x*3) q;
}
gate g2(x) q {
   g3(x/3) q;
}
```
then `g1(5*(pi/3))` evaluates as 
```
tmp_angle0 = (5*(pi/3));
tmp_angle1 = tmp_angle0*3;
tmp_angle2 = tmp_angle1/3;
g3(tmp_angle2);
```
ie `g3(pi/3)`.

### Part 3: what to do about `U`
There are 2 built-in parameterized gates, `U(θ, ϕ, λ)` and `gphase(γ)`. The definitions of these gates are explicitly 2π periodic in all the parameters except for `θ`, for which `U` is instead 4π periodic. The spec implies that `U()` is defined for  
θ in the interval [0,2π), but also says that the `inv` modifier can operate by replacing `inv @ U(θ, ϕ, λ)` by `U(-θ, -λ, -ϕ)`. Unfortunately this is not correct when considering global phase: `ctrl @ inv @ U(θ, 0, 0) c, t;` is not the same as `ctrl @ U(2pi-θ, 0, 0) c, t;` but rather needs an additional `p(pi) c;` to correct.

I can see a few options to resolve this:

1. (not preferred) redefine `U(θ, ϕ, λ)` to be equivalent to `U(θ/2, ϕ, λ)` to that all parameters can be 2pi periodic. This would be a big change to the language and i think we can discount it immediately;
5. redefine angles to overflow mod 4pi rather than mod 2pi. Pro: maintains the definition of `U` and that `inv @ U(θ, ϕ, λ)` is equivalent to `U(-θ, -λ, -ϕ)`. Pro: `U` is continuous and differentiable in all its parameters everywhere. Con: One extra bit needed to store angles. Con: a change to the behaviour of angles (although it's rarely observable and unlikely that many were relying on it)
6. clarify that indeed the θ parameter wraps mod 2pi and that therefore `inv @ U(θ, ϕ, λ)` has an extra `gphase`. Pro: minimal change to the existing spec. Con: `U` gains a discontinuity.
7. redefine the global phase of `U` so that `U(θ, ϕ, λ)` is equivalent to `exp(i(θ+ϕ+λ)/2) Rz(ϕ).Ry(θ).Rz(λ)`. We already changed the global phase going from qasm2 to qasm3, so we could have done this at the time and I wish we thought of it then. Pro: no worse relative to qasm2 than the previous definition. Pro: difference from previous definition only visible in presence of `ctrl` modifier. Pro: U gate is everywhere continuous and differentiable. Con: a difference from the previous spec version and explicit U definition in the QASM3 paper.

I think any of the options 2-4 could be acceptable but I lean towards option 4 even though its a bit embarrassing to have to revisit this and obsolete the paper.

### Part 4: what about precision?
There's a need to choose a number of digits for the calculations. The concrete parameter of a gate call will have a specific precision, and the defcal will have specific precisions, but there is no precision specified for the intermediate steps. I propose to leave the details implementation-defined but with the intention that the general level of precision of these intermediate calculations should be sufficient to realize the defcal precision. One approach might be to do all intermediate calculation at some fixed machine word size (eg the size of `int` without size modifier) and round to the defcal precision just before passing control into the defcal.

Maybe there is some care needed with the precision so that the same rounding always occurs otherwise in the vicinity of the wraparound you could have some bad behaviour:
```
      if 2*a > pi:  // tag1
           xpi q;
           rx(2*a-pi) q;  // tag2
      else:
           rx(2*a) q;
```
if the `a` is close to pi and the `2*a` terms might wrap differently in at `tag1` than at `tag2` then could get an unwanted result.

**Comment by ndebeaudrap** [comment]
I think now, the Types and Casting WG determined that the type of gate parameters are in principle ``angle``s, with a number of bits of precision which may depend on the implementer, and which have wrapping modulo ``2π`` (corresponding in particular to option #3 above in what to do about `U`). Does this seem accurate?

**Comment by mbhealy** [comment]
> I think now, the Types and Casting WG determined that the type of gate parameters are in principle `angle`s, with a number of bits of precision which may depend on the implementer, and which have wrapping modulo `2π` (corresponding in particular to option #3 above in what to do about `U`). Does this seem accurate?

Seems accurate to me.

**Comment by blakejohnson** [comment]
Resolved by #393.
